### ISSUE TITLE
Implement Cross-Platform Kerberos for .NET Core

### ISSUE THREAD

**Issue by alexkeh** [dotnet/corefx#32291]
Kerberos is a popular authentication protocol that many Windows .NET applications use. We would like .NET Core to add a cross-platform Kerberos implementation, including support for these MIT Kerberos APIs or equivalent functionality:

- krb5_init_context((krb5_context *)&context);
- krb5_sname_to_principal(context, (char *) 0, (char *) 0, KRB5_NT_SRV_HST, &server);
- krb5_get_default_realm(context, &therealm);
- krb5_cc_resolve(context, CCNm_, &ccache);
- krb5_cc_get_principal(context, ccache, &client); 
- krb5_auth_con_init(context, &auth_context);
- krb5_copy_principal(context, server, &creds.server);
- krb5_copy_principal(context, client, &creds.client);
- krb5_get_credentials(context, 0, ccache, &creds, &credsp);
- krb5_mk_req_extended(context, &auth_context, auth_options, &cksum_data, credsp, &outbuf);

**Comment by karelz** [comment]
What are the user scenarios taking advantage of the APIs?
Do you have some API shape in mind?
How should it be integrated into existing networking APIs?

cc @wfurt @davidsh

**Comment by ScotMac** [comment]
We would prefer that it is NOT integrated into any existing Network APIs, since that gives the API user more flexibility in terms of how to use or send the data to the server...which is the same as how MIT Kerberos5 APIs do it. Specifically, this is not at all associated with a web server or client.

**Comment by karelz** [comment]
So you are looking for general API for the protocol. Does anything prevent anyone from implementing it as separate library? Why does it need to be in CoreFX / .NET Core? Just convenience?
I assume there are similar APIs in Windows for Kerberos. Is that correct?

**Comment by jkotas** [comment]
https://github.com/SteveSyfuhs/Kerberos.NET ?

**Comment by filipnavara** [comment]
.NET already provides some high-level APIs that use Kerberos functionality internally, such as `NegotiateStream`. It may make sense to expose the low-level APIs in a platform independent way and layer the high level ones on top of them.

We already had to write platform specific versions of the code in order to support Kerberos authentication within the context of SASL authentication, which is in turn used by application protocols such as SMTP, IMAP or LDAP.

Unfortunately the high-level API already provided is not suitable for this use case since we are bound by another protocol standard, where the on-wire representation is different.

**Comment by karelz** [comment]
If there is room for reuse of APIs even inside .NET Core itself, then it would definitely make sense.
We would need someone familiar with those implementations to help design the right API surface - satisfying the consumer needs & making it reasonable for public consumption.

**Comment by ScotMac** [comment]
> .NET already provides some high-level APIs that use Kerberos functionality internally, such as `NegotiateStream`. It may make sense to expose the low-level APIs in a platform independent way and layer the high level ones on top of them.
> 
> We already had to write platform specific versions of the code in order to support Kerberos authentication within the context of SASL authentication, which is in turn used by application protocols such as SMTP, IMAP or LDAP.
> 
> Unfortunately the high-level API already provided is not suitable for this use case since we are bound by another protocol standard, where the on-wire representation is different.

Yes, this is exactly the problem. The windows APIs and .NET APIs are within the context of very specific upper layer protocols, and often a windows domain instead of a generic KDC.

**Comment by ScotMac** [comment]
> https://github.com/SteveSyfuhs/Kerberos.NET ?

Steve's package does no authenticator generation, only validation.

**Comment by ScotMac** [comment]
> So you are looking for general API for the protocol. Does anything prevent anyone from implementing it as separate library? Why does it need to be in CoreFX / .NET Core? Just convenience?
> I assume there are similar APIs in Windows for Kerberos. Is that correct?

Yes, it doesn't have to be in .NET/.NET Core. For example, if Steve's package could be extended to do the GENERATION of the authenticator, that would be a great first step.  

But yes, it would be a lot "nicer" if it was included in .NET/.NET Core.

**Comment by SteveSyfuhs** [comment]
For what it's worth, the library has been extended to do generation (and issuance and what not).

**Comment by tornie2** [comment]
@SteveSyfuhs 

Does that mean we can generate Kerberos tokens in .net Core now ?
I .net 4.8 we could do it with something like this:

        private string KerberosToken(Uri uri)
        {
            // Uri uri = new Uri(url);
            var serviceName = "HTTP/" + uri.Host + (uri.Port != 80 && uri.Port != 443 ? ":" + uri.Port : "");

            var tokenProvider = new KerberosSecurityTokenProvider(serviceName, TokenImpersonationLevel.Impersonation, CredentialCache.DefaultNetworkCredentials);
            var securityToken = tokenProvider.GetToken(TimeSpan.FromMinutes(1)) as KerberosRequestorSecurityToken;
            return Convert.ToBase64String(securityToken.GetRequest());
        }

**Comment by SteveSyfuhs** [comment]
That class is not available in Core AFAIK. It's also very Windows-specific. The library people are referring to above is 3rd party and is an implementation of the Kerberos protocol. It does not call into GSSAPI, which is probably not the most useful for corefx.

**Comment by karelz** [comment]
The APIs seem to be too low-level for us to take into the BCL. We think the APIs won't be used by too many people.
It sounds like a good candidate for independent package. Closing.

**Comment by ScotMac** [comment]
Whereas it is great that Steve's package has been extended to client side, it doesn't help for the reason that this issue was created, since Steve noted they are only available on Windows. ie, the point here is cross platform (Linux and MacOS Core) support for Kerberos client side functionality.

**Comment by SteveSyfuhs** [comment]
@ScotMac I think maybe you misinterpreted what I said. `KerberosSecurityTokenProvider` is Windows-only, and AFAIK doesn't exist in Core in any form. The library *I built* is cross-platform (or at least intends to be though it's not rigorously tested off Windows).

You can more or less replace calls to `KerberosSecurityTokenProvider` with [KerberosClient](https://github.com/SteveSyfuhs/Kerberos.NET#using-the-kerberos-client), but it doesn't rely on GSSAPI so the behavior and usage is different.

**Comment by ScotMac** [comment]
Got it Steve, thanks. 

So, your package now DOES handle client side authenticator generation? Can you please give an example of the generation (sorry if you already have it somewhere else).

**Comment by ScotMac** [comment]
Also, does your package support BOTH framework and core?

**Comment by SteveSyfuhs** [comment]
Yes, you can do client side authenticator generation. It's all wrapped in the `KerberosClient` class. Here's an E2E test that shows how it's done: https://github.com/SteveSyfuhs/Kerberos.NET/blob/master/Tests/Tests.Kerberos.NET/E2EKdcClientTest.cs#L135-L158

The gist of it is:

1. Create a client (don't pass override KDC unless you need to pin it to a particular DC)
2. Authenticate the client to get the TGT
3. Request a service ticket to some SPN for that authenticated user
4. Send the ticket to some service

It's built as netstandard2.0 so that should in principle work for both framework and core.

Here's the thing though. This doesn't have the same level of security as say Kerberos on Windows through SSPI, because it doesn't rely on LSA protecting it's secrets. This changes the security model significantly and you need to consider whether this is something to account for in your build.

**Comment by ScotMac** [comment]
MIT client libraries support BOTH the ability to see/utilize the client authentication via LSA *and* a Kerberos5 specific local credential cache (kinit/klist and okinit/oklist for Oracle). Are you saying that you don't support the LSA portion, and only support a local credential cache file?

**Comment by SteveSyfuhs** [comment]
I expect that MIT is just calling SSPI since Windows doesn't expose a way to access cached tickets directly.

My library doesn't touch LSA at all. The cache is rudimentary and doesn't even use a file.

________________________________

**Comment by ScotMac** [comment]
Sure, they call LSA APIs to acquire the ccache authentication data, after which they use it just like it came from a ccache file. You can read about KFW (KerberosForWindows) and their support for MSLSA here: https://web.mit.edu/kerberos/kfw-4.1/kfw-4.1.html

We would need support for file based, and probably MSLSA, ccache. 

Your implementation strictly uses in memory data for the ccache? If so, is a file based transformation possible? ie, read in the file and supply the resultant byte array to your library?

**Comment by SteveSyfuhs** [comment]
Okay, so that's unsupported on Windows today. The way it queries LSA is not a supported mechanism, is blocked by default, and hard blocked in certain scenarios. SSPI is the only safe way on Windows clients.

The library could be extended easily enough to support a file-based cache. It's not terribly complicated.

________________________________

**Comment by ScotMac** [comment]
Note, MIT's KFWindows *also* has a way of translating MSLSA to the MIT *API* cache. Here is their description (from the link i gave above):

* ms2mit.exe behavior has changed to improve the MSLSA: cache experience for UAC-restricted login sessions on an AD domain that runs ms2mit.exe in login scripts:
  * If the TGT is accessible in the LSA ccache, copy the LSA ccache to the API ccache.
  * Set the registry key for the default ccname to "API:" if the copy occurred, or to "MSLSA:" if it didn't occur.

**Comment by ScotMac** [comment]
Oh, and our MSLSA via MIT Kerberos5 works fine on all Windows versions. ie, is not blocked/hard.blocked.

**Comment by undecided2013** [comment]
Guys I am having a similar issue. Here's the problem I am trying to solve:
1. Running a .net core 3.1 in a container on Openshift (works fine)
2. Using a keytab file I would like to fetch a kerberos token in order to use for a REST call to a Kerberos protected service

Other teams in the firm have this working in Node and in Java, they are basically accessing the MIT libs and using the Krb5.conf and locally stored keytabs for everything.

So far I have attempted to do this with Kerberos.Net and httpclient but the SPNEGO to that REST service fails.
Is there a native .net core way of doing this?

Thanks

**Comment by wfurt** [comment]
.net only wraps krb5 via GSSPI. So it will respect krb5.conf as well as keytab entries. 
It should be certainly possible. look at https://github.com/dotnet/corefx/pull/39933
You can perhaps also look at the test. If that does not help you I would suggest to open a separate issue and provide as many details as you can @undecided2013.

**Comment by undecided2013** [comment]
Thanks, I just looked at the test code, I do not see any tests that utilize keytabs. To be more clear, I only have username, keytab, spn of REST service to be hit and the uri for that service.
